cpmpy/transformations/linearize.py

```diff
@@ -5,190 +5,183 @@
-
-        return linearize_constraint(flatten_constraint(cons_lhs)) + [c1, c2]
-
-    if cpm_expr.name in [">=", "<=", "=="] and \
-            isinstance(cpm_expr.args[0], Operator) and cpm_expr.args[0].name == "mul":
-        lhs, rhs = cpm_expr.args
-        if len(lhs.args) == 2:
-            # multiplication of var and constant
-            if _wsum_should(lhs):
-                return [Comparison(cpm_expr.name, Operator("wsum", _wsum_make(lhs)), rhs)]
-            return [cpm_expr]
+        # BV -> LinExpr
+        if isinstance(cond, _BoolVarImpl):
+            lin_sub = linearize_constraint(sub_expr, supported=supported)
+            return [cond.implies(lin) for lin in lin_sub]
+        # # LinExpr -> BV Updated: unsupported by new grammar
+        # lin_cond = linearize_constraint(cond, supported=supported)
+        # if len(lin_cond) > 1:
+        #     new_vars, cons = zip(*[get_or_make_var(l_expr) for l_expr in lin_cond])
+        #     return [sum(new_vars) == len(new_vars)] + [cond.implies(sub_expr) for cond in lin_cond]
 
+    # comparisons
+    if isinstance(cpm_expr, Comparison):
         lhs, rhs = cpm_expr.args
-        var, constraints = get_or_make_var(lhs.args[0] * lhs.args[1])
-        # decompose long multiplication in several multiplications with two args
-        for arg in lhs.args[2:]:
-            var, cons = get_or_make_var(var * arg)
-            constraints += cons
 
-        return constraints + [Comparison(cpm_expr.name, var, rhs)]
+        # linearize unsupported operators
+        if lhs.name not in supported: # TODO: add mul, (abs?), (mod?), (pow?)
+            if isinstance(lhs, _NumVarImpl) and is_num(rhs):
+                pass
+            elif isinstance(lhs, _NumVarImpl) and isinstance(rhs, _NumVarImpl):
+                lhs, rhs = lhs + -1 * rhs, 0  # bring numvar to lhs
+
+            elif lhs.name == "sub":
+                # convert to wsum
+                lhs = sum([1 * lhs.args[0] + -1 * lhs.args[1]])
+
+            elif lhs.name == "element":
+                arr, idx = cpm_expr.args[0].args
+                # Assuming 1-d array
+                assert len(arr.shape) == 1, f"Only support 1-d element constraints, not {cpm_expr} which has shape {cpm_expr.shape}"
+
+                n = len(arr)
+                sigma = boolvar(shape=n)
+
+                constraints = [sum(sigma) == 1]
+                constraints += [sum(np.arange(n) * sigma) == idx]
+                # translation with implication:
+                constraints += [s.implies(Comparison(cpm_expr.name, a, cpm_expr.args[1])) for s, a in zip(sigma, arr)]
+                return linearize_constraint(constraints, supported=supported)
+
+            # other global constraints
+            elif isinstance(lhs, GlobalConstraint) and hasattr(lhs, "decompose_comparison"):
+                decomp = lhs.decompose_comparison(cpm_expr.name, rhs)
+                return linearize_constraint(flatten_constraint(decomp),supported=supported)
```

Comment:
Do we not assume everything unsupported is already flattened? Otherwise, could, e.g., a `sum` expression also have nested terms?